avm/src/daemon/daemon.rs

```diff
@@ -173,43 +206,112 @@ pub async fn start(
+            let ips = vms
+              .iter()
+              .map(|vm| vm.private_ip_addr.to_string())
+              .collect();
+            let lrh = LogRendezvousHash::new(ips);
+            leader_ip = lrh.get_leader_id().to_string();
+          }
+          if leader_ip == self_ip {
+            let mut factor_err: Option<String> = None;
+            let factor = post_v1_stats(&cluster_id, &deploy_token)
+              .await
+              .unwrap_or_else(|err| {
+                factor_err = Some(format!("{}", err));
+                return "1".to_string();
+              });
+            if let Some(err) = factor_err {
+              error!(115, "{}", err).await;
+            };
+            println!(
+              "VM stats sent for cluster {} of size {}. Cluster factor: {}.",
+              cluster_id,
+              vms.len(),
+              factor
+            );
+            if factor != "1" {
+              post_v1_scale(&cluster_id, &agzb64, &deploy_token, &factor).await;
+            }
+          }
+        }
+      }
+      (Err(dns_err), Ok(_self_ip)) => {
+        error!(117, "DNS error: {}", dns_err).await;
+        panic!("DNS error: {}", dns_err);
+      }
+      (Ok(_dns), Err(self_ip_err)) => {
+        error!(118, "Private ip error: {}", self_ip_err).await;
+        panic!("Private ip error: {}", self_ip_err);
       }
-      if leader_ip == self_ip {
-        let factor = post_v1_stats(&cluster_id, &deploy_token).await;
-        println!(
-          "VM stats sent for cluster {} of size {}. Cluster factor: {}.",
-          cluster_id,
-          vms.len(),
-          factor
+      (Err(dns_err), Err(self_ip_err)) => {
+        error!(
+          119,
+          "DNS error: {} and Private ip error: {}", dns_err, self_ip_err
+        )
+        .await;
+        panic!(
+          "DNS error: {} and Private ip error: {}",
+          dns_err, self_ip_err
         );
-        if factor != "1" {
-          post_v1_scale(&cluster_id, &agzb64, &deploy_token, &factor).await;
-        }
       }
     }
   });
-  run_agz_b64(agz_b64, priv_key_b64, cert_b64).await;
+  let clone_agz_b64 = String::from(agz_b64);
+  let clone_priv_key_b64: Option<String>;
+  let clone_cert_b64: Option<String>;
+  if let (Some(priv_key), Some(cert)) = (priv_key_b64, cert_b64) {
+    clone_priv_key_b64 = Some(String::from(priv_key));
+    clone_cert_b64 = Some(String::from(cert));
+  } else {
+    clone_priv_key_b64 = None;
+    clone_cert_b64 = None;
+  }
+  let run_agz = task::spawn(async {
+    run_agz_b64(clone_agz_b64, clone_priv_key_b64, clone_cert_b64).await;
+  })
+  .await;
+  match run_agz {
+    Ok(_) => println!("ok"),
+    Err(e) => {
+      println!("{:?}", e);
+      if let Ok(reason) = e.try_into_panic() {
+        match reason.downcast_ref::<&str>() {
+          Some(err) => {
+            eprintln!("{}", err);
+          }
+          None => {
+            eprintln!("{:?}", reason);
+          }
+        }
+      }
```

Comment:
What is this whole block about?

Reply:
Following [this](https://github.com/alantech/alan/pull/468#discussion_r607050426), as I'm spawning a new task to run the agz file and if something panics I'm catching it in this block, if the reason of the `JoinError` returned by the spawned task was a panic, get the ref and be able to print/log the string error message.